### ISSUE TITLE
ossl_decoder_fast_is_a is relatively heavy-weight

### ISSUE THREAD

**Issue by beldmit** [openssl/openssl#20399]
Follow-up to #20286 

The code from https://gist.github.com/richsalz/31a98a3095fa36ab6a66082a0c028a6f decodes 65536 Ed25519 private keys in a loop. It makes ~153 mln iterations of ossl_decoder_fast_is_a consuming 20% of time (with descendants) - 2300 iterations per key. It looks suspicious to me.

There are several more functions called 120-160 mln times but they seem to consume 0.5-1% of time each (with descendants).

The callgrind log is attached.

[callgrind.out.655458.txt](https://github.com/openssl/openssl/files/10849028/callgrind.out.655458.txt)

**Comment by beldmit** [comment]
@openssl/otc a follow-up to today's meeting

**Comment by slontis** [comment]
The supplied code is doing a decode and an encode..

I had a look at debugging the decode part with the following c code 
````
static const char private_header[] = {
    0x30, 0x2e,          /* SEQUENCE(3 elem) (46 bytes)     */
    0x02, 0x01, 0x00,    /*   INTEGER 0                     */
    0x30, 0x05,          /*   SEQUENCE(1 elem) (5 bytes)    */
    0x06, 0x03,          /*     OBJECT IDENTIFIER (3 bytes) */
    0x2b, 0x65, 0x6e,    /*       OBJ_X25519 -> 1.3.101.110 -> 1*40+3, 101, 110 */
    0x04, 0x22,          /*   OCTET STRING(1 elem) (nested) */
    0x04, 0x20           /*     OCTET STRING(32 bytes)      */
};

static int dopkcs8_decode(const unsigned char *privdata, size_t privdata_len)
{
    int ret = 0;
    unsigned char der[64];
    const unsigned char *pder = der;
    long derlen;
    PKCS8_PRIV_KEY_INFO *p8info = NULL;
    EVP_PKEY *key = NULL;
    size_t sz = sizeof(private_header);

    memcpy(der, private_header, sz);
    memcpy(der + sz, privdata, privdata_len);
    sz += privdata_len;
    derlen = sz;

    p8info = d2i_PKCS8_PRIV_KEY_INFO(NULL, &pder, derlen);
    if (p8info == NULL)
        goto err;

    key = EVP_PKCS82PKEY_ex(p8info, mainctx, NULL);
    if (key == NULL)
        goto err;
/*
  X509_PUBKEY *ppubkey = nullptr;
  if (1 != X509_PUBKEY_set(&ppubkey, key))
  {
    return nullptr;
  }

  const unsigned char *p = nullptr;
  int plen;

  if (1 != X509_PUBKEY_get0_param(nullptr, &p, &plen, nullptr, ppubkey) || !p || plen != k25519PublicKeyBytes)
  {
  }
  else
  {
  }

  X509_PUBKEY_free(ppubkey);
*/
    ret = 1;
err:
    EVP_PKEY_free(key);
    PKCS8_PRIV_KEY_INFO_free(p8info);
    return ret;
}

static int test_pkcs8_decode(void)
{
    unsigned char i, j;
    unsigned char private[32] = { 0 };

    for (i = 0; i < 1; i++) {
        private[0] = i;
        for (j = 0 ; j < 1; j++) {
            private[1] = j;
            dopkcs8_decode(private, sizeof(private));
        }
    }
    return 1;
}
````

I am thinking that trying to 'optimize' the decoder may be the wrong approach.
This particular code collects 18 keymanagers and 22 decoders
The 17th decoder is X25519 so that is a lot of decoders to run thru get to the right one.
It seems like decoding the blob to figure out what the decoder type, and OID is may be the best approach,
and then using that as a hint. And then falling back in the case where if the data is unknown.

What is particularly bad about this particular case is that there is a function in the legacy code that extracts the OID :)
````
    if (!PKCS8_pkey_get0(&algoid, NULL, NULL, NULL, p8))
        return NULL;
````
So couldn't it at least do that (that doesnt fix the general problem, but it addresses this one).
And use it to figure out the keytype passed to OSSL_DECODER_CTX_new_for_pkey()

**Comment by slontis** [comment]
@beldmit Was that a debug build for the callgrind log? (The numbers I measured in another test were quite different in release)

**Comment by levitte** [comment]
@slontis, when saying "It seems like decoding the blob to figure out what the decoder type, and OID is may be the best approach", I wonder in what part this should happen...

**Comment by beldmit** [comment]
@slontis 
```compiler: gcc -fPIC -pthread -m64 -Wa,--noexecstack -Wall -O3 -ggdb -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DNDEBUG```

**Comment by beldmit** [comment]
I believe that on different machines the result may be different

**Comment by mattcaswell** [comment]
> @slontis, when saying "It seems like decoding the blob to figure out what the decoder type, and OID is may be the best approach", I wonder in what part this should happen...

I think the suggestion was to do that in `EVP_PKCS82PKEY_ex` prior to calling `OSSL_DECODER_CTX_new_for_pkey` - which helps in this very specific case.

**Comment by beldmit** [comment]
@slontis sorry but how do those 18+22 turn into 2300 calls?

**Comment by levitte** [comment]
> > @slontis, when saying "It seems like decoding the blob to figure out what the decoder type, and OID is may be the best approach", I wonder in what part this should happen...
> 
> I think the suggestion was to do that in `EVP_PKCS82PKEY_ex` prior to calling `OSSL_DECODER_CTX_new_for_pkey` - which helps in this very specific case.

Yeah actually, that's not a dumb idea.  If the OID is converted to name or canonical text form (`OBJ_obj2txt()` does that), we should probably be able to use that directly as the keytype, i.e. replace `NULL` with that in this line:

https://github.com/openssl/openssl/blob/d0a3b9d1eb1fc510ec3447b44803bbf5520a0c47/crypto/evp/evp_pkey.c#L84-L85

**Comment by mattcaswell** [comment]
I'm working on it

**Comment by mattcaswell** [comment]
See #20412. I ran @slontis's code (based on original code from @richsalz) but modified the loops to do 256x256 iterations as per the original. I saw these timings on my system:

1.1.1:
real	0m6.791s
user	0m6.783s
sys	0m0.009s

3.0:
real	0m47.326s
user	0m42.134s
sys	0m0.001s

3.1:
real	0m14.663s
user	0m14.658s
sys	0m0.005s

master:
real	0m15.105s
user	0m15.091s
sys	0m0.013s

master with #20412:
real	0m8.887s
user	0m8.887s
sys	0m0.000s

So the patch approximately doubles performance on master.

**Comment by mattcaswell** [comment]
Obviously this is only a partial fix for the issue because it only really improves the `EVP_PKCS82PKEY_ex`/`EVP_PKCS82PKEY` functions and not decoders in general.

**Comment by mattcaswell** [comment]
1.1.1 -> master with #20412 is about 30% slower. Not fantastic but significantly better than the 216% slower for 1.1.1 -> 3.1 currently or 697% for 1.1.1 -> 3.0

**Comment by beldmit** [comment]
Nice work! I confirm 2-times speedup

**Comment by beldmit** [comment]
[callgrind.out.15039.txt](https://github.com/openssl/openssl/files/10860900/callgrind.out.15039.txt)
Callgrind after this patch looks much better.

**Comment by beldmit** [comment]
> 

I'm not sure it's realistic to speedup the whole encoder/decoder system, but there are some low-hanging fruits to cut when we find them.

**Comment by mattcaswell** [comment]
See also #20416. That's still only a partial fix, but it perhaps impacts more widely than #20412.

**Comment by paulidale** [comment]
I like @mattcaswell's approach of probing for well known formats.  We might want to think about how to do it more generically (read a block of data, try to guess type, try that/those first & fall back to what we've got now).  It needn't be perfect, just better for common cases.

**Comment by paulidale** [comment]
A probe into a PEM file would see the _first_ item and might provide a hint as to what it contains (or not).  A PEM file with multiple parts would be missed by the probe but picked up later in the fall back.  I.e. it shouldn't be worse unless we retry "failed" deciders.

**Comment by slontis** [comment]
Yeah but it still doesnt know the keytype for some of these.

**Comment by richsalz** [comment]
Does OpenSSL ever parse multiple items in PEM format that are in a single file? Isn't that up to the application to do? (I could well be out of date here.)

Seems to me that if you have the PEM header you have a fast-path to look at the likely decoder. And you could always decode the base64 contents, and look to see if the DER agrees with the label, and raise an error if not.

**Comment by slontis** [comment]
Pretty sure if handles multiple items.. That would have to fallback.
-----BEGIN PRIVATE KEY------ doesnt completely help :)

**Comment by richsalz** [comment]
Yes, generic tags would mean more work.  For example, unbase64 and look at the DER.

> Pretty sure it handles multiple items.

Can you point to an API in libcrypto that does this?

**Comment by slontis** [comment]
Might be a question for @levitte..

**Comment by levitte** [comment]
Regarding PEM, the PEM functions that take a BIO or file descriptor will parse the next PEM object in that stream, so libcrypto at least *supports* reading multiple objects in an iterative fashion.  The OSSL_STORE API supports the same thing, obviously, but for any sort of container (not just PEM files).

It's true, though, that it's up to the application to perform the actual iteration over multiple objects in a container, generally speaking.

One might argue that OpenSSL supports reading multiple objects in one go, though, when parsing a PKCS#12 object.  You do get certs and key in one call (more or less).  But, that's not about PEM.

**Comment by beldmit** [comment]
If I understand correctly, CAfile argument of command line utilities may contain several certificates

**Comment by levitte** [comment]
I think the big question here is, how to detect what chain there should be, with minimal parsing.

Like someone else already said, this is currently a bit hard to do, because we build up all the potentially viable chains before even looking at the data stream.  However, it's conceptually possible to leave that build-up until the first read of the data stream.  Simply leave it to `OSSL_DECODER_CTX_new_for_pkey()` just store its inputs the the resulting `OSSL_DECODER_CTX`, and on the first `OSSL_DECODER_from_` call for that `OSSL_DECODER_CTX`, do some light parsing, put the resulting bytes through a process like the `file` command does, and use that to try to determine what a viable chain should look like.

Note that there are critical stumbling blocks with the proposed analysis.  We don't know what decoders there will be out there, so we will need to be careful about what assumptions we make.

**Comment by mattcaswell** [comment]
> However, it's conceptually possible to leave that build-up until the first read of the data stream. Simply leave it to OSSL_DECODER_CTX_new_for_pkey() just store its inputs the the resulting OSSL_DECODER_CTX, and on the first OSSL_DECODER_from_ call for that OSSL_DECODER_CTX, do some light parsing, put the resulting bytes through a process like the file command does (heh, does that mean we should learn to use /etc/magic?), and use that to try to determine what a viable chain should look like.

I actually did some experiments with this same idea yesterday, but I couldn't get it working properly in the time available. I do think this sounds like a potentially viable approach though.

**Comment by levitte** [comment]
> I actually did some experiments with this same idea yesterday, but I couldn't get it working properly in the time available. I do think this sounds like a potentially viable approach though.

I'm guessing that this is mostly relevant when the input is raw DER, as such files *usually* contain only one object (there's one well known exception).  Since PEM files can potentially contain multiple object of all types, all you can be certain of is that it's a PEM file.

**Comment by mattcaswell** [comment]
> I'm guessing that this is mostly relevant when the input is raw DER,

Yes, I agree. We're mostly talking about DER here.

**Comment by richsalz** [comment]
So we have an answer to one question: nothing in the OpenSSL *libraries* read more than one PEM object at a time. They only read forward to get the next one.

>  One might argue that OpenSSL supports reading multiple objects in one go, though, when parsing a PKCS#12 object. You do get certs and key in one call (more or less). But, that's not about PEM.

Well, there are many single-DER objects that are containers (PKCS7, OCSP response) but as you say that's not relevant here. But what did you mean by "one well-known exception?"

**Comment by levitte** [comment]
> So we have an answer to one question: nothing in the OpenSSL _libraries_ read more than one PEM object at a time. They only read forward to get the next one.
> 
> > One might argue that OpenSSL supports reading multiple objects in one go, though, when parsing a PKCS#12 object. You do get certs and key in one call (more or less). But, that's not about PEM.
> 
> Well, there are many single-DER objects that are containers (PKCS7, OCSP response) but as you say that's not relevant here. But what did you mean by "one well-known exception?"

X509_AUX, which is a X509 DER blob directly followed by an auxilliary data blob.  This was made years (eons? :wink:) ago.

See https://github.com/openssl/openssl/blob/master/crypto/x509/x_x509a.c, and for the parsing itself, this:

https://github.com/openssl/openssl/blob/c4cb151b4d19f5f2eb302baafac65e36e4408381/crypto/x509/x_x509.c#L183-L218

**Comment by richsalz** [comment]
Thank you, @levitte.

**Comment by nhorman** [comment]
@mattcaswell is there more work to be done here?  It seems like you've fixed some problems, but I can't determine if there are concrete subsequent issues to address

**Comment by mattcaswell** [comment]
I think mostly what remains falls into the "decoders are slow" bucket for which we have other issues (e.g. https://github.com/openssl/project/issues/100).

We might also want to consider whether the reproducer supplied here is adequately reflected in our existing performance test suite - or do we need an additional test. We have https://github.com/openssl/tools/blob/master/perf/pkeyread.c already - does this add something sufficiently new? Requires some investigation.

**Comment by nhorman** [comment]
Ok, so I'm going to mark this as consideration for next minor release, and we can schedule that work when available
